velox/connectors/hive/S3Config.h

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <optional>
+#include <string>
+#include "velox/common/base/Exceptions.h"
+
+namespace facebook::velox::config {
+class ConfigBase;
+}
+
+namespace facebook::velox::connector::hive {
+
+/// Build config required to initialize an S3FileSystem instance.
+/// All hive.s3 options can be set on a per-bucket basis.
+/// The bucket-specific option is set by replacing the hive.s3. prefix on an
+/// option with hive.s3.bucket.BUCKETNAME., where BUCKETNAME is the name of the
+/// bucket.
+/// When connecting to a bucket, all options explicitly set will override the
+/// base hive.s3. values.
+/// These semantics are similar to the Apache Hadoop-Aws module.
+/// https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html
+class S3Config {
+ public:
+  S3Config() = delete;
+
+  /// S3 config prefix.
+  static constexpr const char* kS3Prefix = "hive.s3.";
+
+  /// S3 bucket config prefix
+  static constexpr const char* kS3BucketPrefix = "hive.s3.bucket.";
+
+  /// Log granularity of AWS C++ SDK.
+  static constexpr const char* kS3LogLevel = "hive.s3.log-level";
+
+  /// S3FileSystem default identity.
+  static constexpr const char* kDefaultS3Identity = "s3-default-identity";
+
+  /// Keys to identify the config.
+  enum class Keys {
+    kBegin,
+    kEndpoint = kBegin,
+    kAccessKey,
+    kSecretKey,
+    kPathStyleAccess,
+    kSSLEnabled,
+    kUseInstanceCredentials,
+    kIamRole,
+    kIamRoleSessionName,
+    kConnectTimeout,
+    kSocketTimeout,
+    kMaxConnections,
+    kMaxAttempts,
+    kRetryMode,
+    kUseProxyFromEnv,
+    kEnd
+  };
+
+  /// Map of keys -> <suffixString, optional defaultValue>.
+  /// New config must be added here along with a getter function below.
```

Comment:
maybe give it a name `using TConfigMap = ...`

Reply:
done